### PR TITLE
Basic features

### PR DESCRIPTION
**Pull request by insxne**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,6 +14,9 @@
   "license": "UNLICENSED",
   "private": true,
   "dependencies": {
-    "hapi": "^11.0.2"
+    "async": "^1.4.2",
```
Comment (joaquinrulin) [inline package.json]: @insxne delete the unused async library plz

File: lib/persistence/redis.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    const keys = this.client.keys(search);
+    const values =  keys.then((keys) => Promise.all(
+      keys.map((key) => this.client.get(key))
+    ));
+    return Promise.all([keys, values]).then((keysValues) => {
+      return Redis.object(keysValues[0], keysValues[1])
+    });
+  }
+
+  static object(list, values) {
```
Comment (joaquinrulin) [inline lib/persistence/redis.js]: @insxne may you use underscore's namesake method?

File: lib/.jshintrc [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (joaquinrulin) [inline lib/.jshintrc]: @insxne as we are gonna use eshint better delete this

File: index.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
+const Hapi = require('hapi');
+const Joi = require('joi');
+const Text = require('./lib/text');
+
+// Create a server with a host and port
+var server = new Hapi.Server();
+server.connection({
+  host: 'localhost',
```
Comment (joaquinrulin) [inline index.js]: @insxne create a config file for storing both host and port

**Comment by warseph** [comment]
Why is the `dump.rdb` binary file added?

File: index.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  method: 'GET',
+  path:'/save',
+  handler: function (request, reply) {
+    reply(new Text(request.query).save())
+  },
+  config: {
+    validate: {
+      query: {
+        application: Joi.string().required(),
+        countryCode: Joi.string().required(),
```
Comment (warseph) [inline index.js]: It would be nice to add some more validations, such as length.

File: lib/.jshintrc (hunk 3 above)
Comment (warseph) [inline lib/.jshintrc]: Why are we gonna use eshint @joaquinrulin ?

Comment (joaquinrulin) [inline lib/.jshintrc]: @insxne don't know, @insxne told me is more updated than jshint and i believed him

File: lib/persistence.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+var Redis = require('./persistence/redis');
+
+class Persistence {
+  static factory(){
+    return new Redis();
+  }
+}
```
Comment (warseph) [inline lib/persistence.js]: Is this better than doing just something like this?
```javascript
module.exports = { 
  factory: () => new Redis()
};
```

File: lib/persistence.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+var Redis = require('./persistence/redis');
```
Comment (warseph) [inline lib/persistence.js]: This should be a `const` instead of a `var`

File: index.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
+const Hapi = require('hapi');
+const Joi = require('joi');
+const Text = require('./lib/text');
+
+// Create a server with a host and port
+var server = new Hapi.Server();
```
Comment (warseph) [inline index.js]: This should be a `const` instead of a `var`

File: lib/persistence/redis.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    var result = {};
+    for (var i = 0, length = list.length; i < length; i++) {
```
Comment (warseph) [inline lib/persistence/redis.js]: This should be a `let` instead of a `var`

File: lib/persistence/redis.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    const values =  keys.then((keys) => Promise.all(
+      keys.map((key) => this.client.get(key))
+    ));
+    return Promise.all([keys, values]).then((keysValues) => {
+      return Redis.object(keysValues[0], keysValues[1])
+    });
+  }
+
+  static object(list, values) {
+    var result = {};
```
Comment (warseph) [inline lib/persistence/redis.js]: This should be a `const` instead of a `var`

File: lib/text.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  search() {
+    return persistenceClient.findAll(this.persistenceKey())
+  }
+
+  persistenceKey(params){
+    params = params || this.attributes;
+
+    let keyArray = [
```
Comment (warseph) [inline lib/text.js]: This should be a `const` instead of a `let`

File: lib/text.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+      params.application,
+      params.countryCode,
+      params.url,
+      params.key
+    ];
+    let normalizedKeyArray = keyArray.map(k => k ? k : '*');
```
Comment (warseph) [inline lib/text.js]: This should be a `const` instead of a `let`

File: lib/persistence/redis.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+      return result;
+    });
+  }
+
+  find(key) {
+    const getPromise = this.client.get(key);
+    return Promise.all([key, getPromise]).then((keyValue) => {
+      const result = {};
+      result[keyValue[0]] = keyValue[1];
+      return result;
```
Comment (warseph) [inline lib/persistence/redis.js]: You should extract this object creation for a function (and also use it in save), something like:
```javascript
function getObject(key, value) {
  const result = {};
  result[key] = value;
  return result;
}
```

File: lib/persistence/redis.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
+const RedisLib = require('promise-redis')();
+const Persistence = require('../persistence');
+
+class Redis {
+  constructor() {
+    this.client = RedisLib.createClient();
+    this.client.on("error", function (err) {
+      console.log("Error: " + err);
```
Comment (warseph) [inline lib/persistence/redis.js]: Just logging is enough? can it proceed without a client? shouldn't you retry or stop the program if it fails?

File: lib/text.js [hunk 15]
```diff
@@ -0,0 +1,40 @@
+const Persistence = require('./persistence');
+const persistenceClient = Persistence.factory();
```
Comment (warseph) [inline lib/text.js]: Maybe to be consistent with your other modules you could just do:
```javascript
const persistenceClient = require('./persistence').factory();
```

File: test.js [hunk 16]
```diff
@@ -0,0 +1,11 @@
+class Text {
+  constructor(value, attributes) {
+    attributes = attributes || {};
+    this.value = value;
+    this.attributes = attributes;
+
+    console.log(`value:${value} attributes:${attributes}`);
```
Comment (warseph) [inline test.js]: This looks like a debug, should you really be logging this?

**Comment by warseph** [comment]
As a more general comment, I don't quite get the Persistence Factory class.
From what I can see, you just need one persistence object in the application, and as node isn't usually coded using Dependency Injection, it doesn't require having factories.
I would understand using a Persistence singleton class from which you can get the persistence, which would allow you to switch the object in testing, for example.

Still all this changes can be done at a later stage, It's usually a better practice to not over-architect the application at these early stages. I would just remove the factory altogether and just create the new Redis object where it's needed.

**Pull request merged**